ESOP minimization

SmaranTum opened the pull request:
## Description

This PR includes the addition of an ESOP minimization function (optimization category) which helps in reducing the overall gate count by removing redundant gates.

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [x] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

burgholzer commented on src/algorithms/synthesis/dd_synthesis.cpp:
```diff
@@ -133,19 +133,39 @@ namespace syrec {
+    }
+
+    // This function returns the ESOP minimized form of a signature.
+    auto DDSynthesizer::esopOptimization(TruthTable::Cube::Vector const& sigVec) -> std::vector<minbool::MinTerm> {
+        std::vector<uint64_t> on;
+        on.reserve(sigVec.size());
+
+        for (auto const& sigVecToInt: sigVec) {
+            if (!sigVecToInt.empty()) {
+                on.emplace_back(sigVecToInt.toInteger());
             }
         }
+        return minbool::minimizeBoolean(on, sigVec[0].size());
```
Why does this function not handle don't care conditions? In anticipation of the future this would make a lot of sense and is definitely supported by the underlying library.

SmaranTum replied:
Currently, the don't care conditions are not required. Also, for the future based on the technique that I employ in the synthesis technique, I don't think we will require this as well. I have removed some statements and functions that are concerning the `dcValues` since it is not required. 

if dc conditions are inevitably required, I will add those statements and functions back in the future commits.

burgholzer replied:
Sounds good to me. The removal of dc handling should make some of the functions in the minimization a lot easier.

burgholzer commented on src/algorithms/synthesis/dd_synthesis.cpp:
```diff
@@ -169,13 +189,21 @@ namespace syrec {
-            for (auto const& rootVec: rootSigVec) {
+
+            if (rootSolution.empty()) {
+                dd::Controls ctrlFinal1;
+                applyOperation(nQubits, current.p->v, src, ctrlFinal1, dd);
```
```suggestion
                applyOperation(nQubits, current.p->v, src, {}, dd);
```
Not 100% sure if this works though.

burgholzer commented on src/algorithms/synthesis/dd_synthesis.cpp:
```diff
@@ -259,23 +297,36 @@ namespace syrec {
         controlNonRoot(current, ctrlNonRoot, ctrlVec);
 
         TruthTable::Cube::Vector rootSigVec;
-        pathFromSrcDst(src, current, rootSigVec);
+        pathFromSrcDst(src, current.p, rootSigVec);
 
         const auto nQubits = static_cast<dd::QubitCount>(src.p->v + 1);
-        for (auto const& rootVec: rootSigVec) {
+
+        const auto rootSolution = esopOptimization(rootSigVec);
+
+        if (rootSolution.empty()) {
+            ctrlNonRoot.emplace(dd::Control{current.p->v, dd::Control::Type::pos});
+            for (std::size_t i = 0; i < targetVec.size(); i++) {
+                if (targetVec[i].has_value() && *(targetVec[i])) {
+                    applyOperation(nQubits, static_cast<dd::Qubit>(static_cast<std::size_t>(current.p->v) - (i + 1U)), src, ctrlNonRoot, dd);
+                }
+            }
+        }
+
+        for (auto const& rootVec: rootSolution) {
             dd::Controls ctrlFinal;
-            controlRoot(current, ctrlFinal, rootVec);
+
+            controlRootEsop(current, ctrlFinal, rootVec, rootSigVec[0].size());
+
             ctrlFinal.emplace(dd::Control{current.p->v, dd::Control::Type::pos});
+
             ctrlFinal.insert(ctrlNonRoot.begin(), ctrlNonRoot.end());
 
-            const auto targetSize = targetVec.size();
-            for (auto i = 0U; i < targetSize; ++i) {
+            for (std::size_t i = 0; i < targetVec.size(); i++) {
```
This change should be reverted. Calculating the loop termination condition outside the loop is always preferable to re-compute it in every loop iteration. (Although in cases like these the compiler is probably clever enough to figure that out itself)

burgholzer reviewed (changes requested):
I took quite some time to think about whether adapting an external library is really the best way to do this and about how the template code was adapted. In the end, I believe this is probably the way to go.

Nevertheless, I am not quite happy with the current state. I believe that the external library can be adapted more to fit into our framework in a nicer fashion.
In particular, I think that many changes in the synthesis file are actually not necessary if the minimization library is properly adjusted.

You can find detailed comments down below.

burgholzer commented:
Might have screwed up something during the refactor in the last two commits. Unfortunately, I don't really have time until Tuesday to work further on this. Feel free to revert parts of my commits above.
The main idea was to use proper types (prefixed with `std::` and `const` in a couple more places) and eliminate the need for passing ˋnˋ in many functions.

SmaranTum commented:
> Might have screwed up something during the refactor in the last two commits. Unfortunately, I don't really have time until Tuesday to work further on this. Feel free to revert parts of my commits above. The main idea was to use proper types (prefixed with `std::` and `const` in a couple more places) and eliminate the need for passing ˋnˋ in many functions.

So the error is occurring when explicit is added next to the Minterm constructor. When I remove it works completely fine (That is why while creating the PR I had added a NOLINT thing). Do you may be know why that is happening? I will incorporate the other necessary changes starting from tomorrow.

burgholzer commented:
> > Might have screwed up something during the refactor in the last two commits. Unfortunately, I don't really have time until Tuesday to work further on this. Feel free to revert parts of my commits above. The main idea was to use proper types (prefixed with `std::` and `const` in a couple more places) and eliminate the need for passing ˋnˋ in many functions.
> 
> So the error is occurring when explicit is added next to the Minterm constructor. When I remove it works completely fine (That is why while creating the PR I had added a NOLINT thing). Do you may be know why that is happening? I will incorporate the other necessary changes starting from tomorrow.

Hm. Strange... The only line I changed after introducing the `explicit` was in the return statement of the `combine` method.
That constructor should be explicit and I can't really see a reason why that should be a problem. Might have to dig a bit deeper into that.

SmaranTum commented on include/algorithms/optimization/esop_minimization.hpp:
```diff
@@ -0,0 +1,350 @@
+    }
+
+    struct MinTerm {
+        struct Hash {
+            std::size_t operator()(const MinTerm& term) const {
+                // Bernstein's hash function
+                return 33 * term.value ^ term.dash;
+            }
+        };
```
is it okay if I do like this

namespace minbool{

define Minterm struct

}

namespace std{
//define the necessary hash function
}

//continue with namespace minbool
namespace minbool{}

burgholzer replied:
Yeah. Otherwise you can move the `std` block to the very end of the file.

SmaranTum replied:
If I move the std block to end, it will not work right?  unordered_maps are declared beforehand within the other structures.

burgholzer replied:
That still works. C++ is clever enough to find that definition as long as it exists.

burgholzer commented on include/algorithms/optimization/esop_minimization.hpp:
```diff
@@ -0,0 +1,350 @@
+                std::sort(second.begin(), second.end());
+            }
+
+            for (const auto& [pair1First, pair1Second]: rows) {
+                for (const auto& [pair2First, pair2Second]: rows) {
+                    if (pair1First == pair2First) {
+                        continue;
+                    }
+                    // Dominated rows are eliminated
+                    if (std::includes(pair1Second.begin(), pair1Second.end(),
+                                      pair2Second.begin(), pair2Second.end())) {
+                        rows.erase(pair2First);
+                        change = true;
+                        break;
+                    }
+                }
+            }
```
True. Probably not worth the effort of replacing both Code snippets with a generic one. Feel free to ignore this comment.

SmaranTum commented:
> > > Might have screwed up something during the refactor in the last two commits. Unfortunately, I don't really have time until Tuesday to work further on this. Feel free to revert parts of my commits above. The main idea was to use proper types (prefixed with `std::` and `const` in a couple more places) and eliminate the need for passing ˋnˋ in many functions.
> > 
> > 
> > So the error is occurring when explicit is added next to the Minterm constructor. When I remove it works completely fine (That is why while creating the PR I had added a NOLINT thing). Do you may be know why that is happening? I will incorporate the other necessary changes starting from tomorrow.
> 
> Hm. Strange... The only line I changed after introducing the `explicit` was in the return statement of the `combine` method. That constructor should be explicit and I can't really see a reason why that should be a problem. Might have to dig a bit deeper into that.

Apart from this issue (due to which the cpp linter is failing), I have incorporated most of the necessary changes.

burgholzer commented:
> > > > Might have screwed up something during the refactor in the last two commits. Unfortunately, I don't really have time until Tuesday to work further on this. Feel free to revert parts of my commits above. The main idea was to use proper types (prefixed with `std::` and `const` in a couple more places) and eliminate the need for passing ˋnˋ in many functions.
> > > 
> > > 
> > > So the error is occurring when explicit is added next to the Minterm constructor. When I remove it works completely fine (That is why while creating the PR I had added a NOLINT thing). Do you may be know why that is happening? I will incorporate the other necessary changes starting from tomorrow.
> > 
> > 
> > Hm. Strange... The only line I changed after introducing the `explicit` was in the return statement of the `combine` method. That constructor should be explicit and I can't really see a reason why that should be a problem. Might have to dig a bit deeper into that.
> 
> Apart from this issue (due to which the cpp linter is failing), I have incorporated most of the necessary changes.

I am not 100% sure what changed, but it seems that some of the other changes actually resolved the segfault issue. Everything seems to work as intended now with the `explicit` included.

SmaranTum commented:
> > > > > Might have screwed up something during the refactor in the last two commits. Unfortunately, I don't really have time until Tuesday to work further on this. Feel free to revert parts of my commits above. The main idea was to use proper types (prefixed with `std::` and `const` in a couple more places) and eliminate the need for passing ˋnˋ in many functions.
> > > > 
> > > > 
> > > > So the error is occurring when explicit is added next to the Minterm constructor. When I remove it works completely fine (That is why while creating the PR I had added a NOLINT thing). Do you may be know why that is happening? I will incorporate the other necessary changes starting from tomorrow.
> > > 
> > > 
> > > Hm. Strange... The only line I changed after introducing the `explicit` was in the return statement of the `combine` method. That constructor should be explicit and I can't really see a reason why that should be a problem. Might have to dig a bit deeper into that.
> > 
> > 
> > Apart from this issue (due to which the cpp linter is failing), I have incorporated most of the necessary changes.
> 
> I am not 100% sure what changed, but it seems that some of the other changes actually resolved the segfault issue. Everything seems to work as intended now with the `explicit` included.

Oh wow. it works, that's important.
